click_odoo_contrib/update.py

```diff
@@ -88,13 +198,46 @@ def OdooEnvironmentWithUpdate(database, ctx, **kwargs):
                 "Updating addons for their hash changed: %s.",
                 ",".join(to_update.keys()),
             )
-    if ctx.params["i18n_overwrite"]:
-        odoo.tools.config["overwrite_existing_translations"] = True
+    odoo.tools.config["overwrite_existing_translations"] = ctx.params["i18n_overwrite"]
     if odoo.release.version_info[0] < 10:
         Registry = odoo.modules.registry.RegistryManager
     else:
         Registry = odoo.modules.registry.Registry
-    Registry.new(database, update_module=True)
+    try:
+        # Update Odoo spawning a different process
+        update_process = multiprocessing.Process(
+            target=Registry.new, kwargs={"db_name": database, "update_module": True}
+        )
```

Comment:
To be less intrusive, would it be possible to run the watcher in a subprocess (or thread) only if activated by an option, and let the update run in the main process in all cases?

Reply:
I thought about that option, but let me share my thoughts on the matter:

1. [Only the process that spawned another process is able to join or terminate it](https://docs.python.org/2.7/library/multiprocessing.html#multiprocessing.Process.terminate). If, at some point, Odoo is able to properly handle signals while locked and reap its cursors, it'd be even better to terminate this process top-down instead of bottom-up as it is right now (due to this restriction). In such case, we'd have to terminate it from the main process anyways.
1. Handling 2 subprocesses and making one watch over the other is way harder than with only 1 subprocess. At the end of the day, we'd have to:

   - spawn update
   - spawn watcher
   - do a `while True` loop until both finish

   It's more complex than the current, but at the end of the day would do the same. Current:

   - spawn update
   - watch in a `while True` loop
1. I could add an option that avoids the watcher, and in such case we could avoid spawning processes and behave as before, but after all the watcher is supposed to detect 0 locks if we are not updating in parallel, so at the end of the day nobody would notice it if not using parallel updates, and we'd have to maintain 2 ways of using the script instead of just 1 that, although is more complex, covers both scenarios.

Please let me know what you think about it. :blush:

Reply:
I was thinking about something like this:
- launch watcher in a daemon thread
- main process does the update
- when watcher detects deadlocks it terminates the postgres backend (which in turn should raise an exception that terminates the main update process)
- the daemon watcher thread is killed automatically when the main process exits

Reply:
OK, that makes more sense, but which would be the benefits?

Reply:
@Yajo the change is less intrusive (the normal flow stays unchanged) and you can start the watcher thread only when activated with an option. So less risk and the code would be easier to understand (by me at least) and maintain.

Reply:
I don't think we should add an option to trigger the watcher... :thinking: 

If you don't update in parallel (AFAIK you don't do this actually), the watcher will basically do nothing.

If you do update in parallel, then you always need the watcher.

So, what do we gain on adding an option? Isn't it better to have a watcher that just works always?

Reply:
Well, the watcher will kill the update if a query lasts for more than 30 seconds by default. It's an advanced feature that can have unexpected effects, so its better to let advanced users activate it when they understand they need it.

Reply:
> the watcher will kill the update if a query lasts for more than 30 seconds

Actually not. It will kill the update _if there's a query locking another query_ for more than 30 seconds.

A single process would never have 2 queries in parallel, and even less 2 blocking ones, so it should never affect.

This situation can _only_ happen when there are more than 1 clients connected.